presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedStatsCalculator.java

```diff
@@ -92,77 +96,55 @@ public Visitor(Session session, Map<Symbol, Type> types)
         }
 
         @Override
-        protected Void visitPlan(PlanNode node, Map<PlanNode, PlanNodeCost> context)
+        protected PlanNodeStatsEstimate visitPlan(PlanNode node, List<PlanNodeStatsEstimate> sourceCosts)
         {
-            visitChildren(node, context);
-            context.put(node, UNKNOWN_COST);
-            return null;
+            // TODO: Explicitly visit GroupIdNode and throw an IllegalArgumentException
+            // this can only be done once we get rid of the StatelessLookup
+            return UNKNOWN_STATS;
         }
 
         @Override
-        public Void visitOutput(OutputNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeStatsEstimate visitOutput(OutputNode node, List<PlanNodeStatsEstimate> sourceCosts)
         {
-            visitChildren(node, context);
-            context.put(node, context.get(node.getSource()));
-            return null;
-        }
-
-        private void visitChildren(PlanNode node, Map<PlanNode, PlanNodeCost> context)
-        {
-            for (PlanNode source : node.getSources()) {
-                source.accept(this, context);
-            }
+            return getOnlyElement(sourceCosts);
         }
 
         @Override
-        public Void visitFilter(FilterNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeStatsEstimate visitFilter(FilterNode node, List<PlanNodeStatsEstimate> sourceCosts)
         {
-            boolean fullyEnforced = false;
-            if (node.getSource() instanceof TableScanNode) {
-                 fullyEnforced = visitTableScan((TableScanNode) node.getSource(), context, node.getPredicate());
-            }
-            else {
-                visitChildren(node, context);
-            }
-
-            PlanNodeCost sourceCost = context.get(node.getSource());
-            final double filterCoefficient = fullyEnforced ? 1.0 : FILTER_COEFFICIENT;
-            PlanNodeCost filterCost = sourceCost
-                    .mapOutputRowCount(value -> value * filterCoefficient);
-            context.put(node, filterCost);
-            return null;
+            PlanNodeStatsEstimate sourceCost = getOnlyElement(sourceCosts);
+            return sourceCost
+                    .mapOutputRowCount(value -> value * FILTER_COEFFICIENT);
         }
 
         @Override
-        public Void visitProject(ProjectNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeStatsEstimate visitProject(ProjectNode node, List<PlanNodeStatsEstimate> sourceCosts)
         {
-            return copySourceCost(node, context);
+            return getOnlyElement(sourceCosts);
         }
 
         @Override
-        public Void visitJoin(JoinNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeStatsEstimate visitJoin(JoinNode node, List<PlanNodeStatsEstimate> sourceCosts)
```

Comment:
Part of the issue is probably that I'm using the LocalQueryRunner so network costs aren't included.  But either way I would expect: 
`a cross join b join c on a1.=c.1 and b.1 = c.1`
to be more expensive than
`a join c on a.1= c1. join b on c.1 = b.1 `
because the cross join should generate lots of extra data that needs to be processed.  But currently they have the same cost.